File: 1095-find-in-mountain-array/1095-find-in-mountain-array.py
# """
# This is MountainArray's API interface.
# You should not implement it, or speculate about its implementation
# """
#class MountainArray:
#    def get(self, index: int) -> int:
#    def length(self) -> int:

class Solution:
    def findInMountainArray(self, target: int, mountain_arr: 'MountainArray') -> int:
        length = mountain_arr.length()
        
        low, high = 1, length - 2
        while low != high:
            mid = (low + high) >> 1
            curr = mountain_arr.get(mid)
            nxt = mountain_arr.get(mid + 1)
            
            if curr < nxt:
                if curr == target:
                    return mid
                if nxt == target:
                    return mid + 1
                low = mid + 1
            else:
                high = mid
        
        peak_index = low
        
        low, high = 0, peak_index
        while low <= high:
            mid = (low + high) >> 1
            curr = mountain_arr.get(mid)
            
            if curr == target:
                return mid
            if curr < target:
                low = mid + 1
            else:
                high = mid - 1
        
        low, high = peak_index + 1, length - 1
        while low <= high:
            mid = (low + high) >> 1
            curr = mountain_arr.get(mid)
            
            if curr == target:
                return mid
            elif curr > target:
                low = mid + 1
            else:
                high = mid - 1
        
        return -1

File: 1095-find-in-mountain-array/1095-find-in-mountain-array.cpp
/**
 * // This is the MountainArray's API interface.
 * // You should not implement it, or speculate about its implementation
 * class MountainArray {
 *   public:
 *     int get(int index);
 *     int length();
 * };
 */

class Solution {
public:
    int findInMountainArray(int target, MountainArray &mountainArr) {
        unordered_map<int, int> cache;
        int n = mountainArr.length(), low = 1, high = n - 2, curr, next;
        
        while (low != high){
            int mid = low + ((high - low) >> 1);
            if (cache.find(mid) != cache.end())
                curr = cache[mid];
            else{
                curr = mountainArr.get(mid);
                cache[mid] = curr;
            }
            
            if (cache.find(mid + 1) != cache.end())
                next = cache[mid + 1];
            else{
                next = mountainArr.get(mid + 1);
                cache[mid + 1] = next;
            }
            
            if (curr < next)
                low = mid + 1;
            else
                high = mid;
        }
        
        int peakIndex = low;
        low = 0, high = peakIndex;
        while (low <= high){
            int mid = low + ((high - low) >> 1);
            if (cache.find(mid) != cache.end())
                curr = cache[mid];
            else
                curr = mountainArr.get(mid);
            
            if (curr == target)
                return mid;
            if (curr < target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        
        low = peakIndex + 1, high = n - 1;
        while (low <= high){
            int mid = low + ((high - low) >> 1);
            if (cache.find(mid) != cache.end())
                curr = cache[mid];
            else
                curr = mountainArr.get(mid);
            
            if (curr == target)
                return mid;
            if (curr > target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        
        return -1;
    }
};